src/lib/y2storage/package_handler.rb

```diff
@@ -0,0 +1,168 @@
+  # storage-related packages to the set of packages to install during
+  # installation, install small sets of packages in the installed system (for
+  # the expert partitioner) as needed.
+  #
+  class PackageHandler
+    include Yast::Logger
+    include Yast::I18n
+
+    attr_reader :pkg_list
+
+    PROPOSAL_ID = "storage_proposal"
+
+    def initialize
+      textdomain("storage-ng")
+      reset
+    end
+
+    # Clear the package list
+    #
+    def reset
+      @pkg_list = []
+    end
+
+    # Add a number of packages to the list of packages to be installed
+    #
+    # @param  [Array<String>] package names
+    # @return [Array<String>] new package list (may contain duplicates)
+    #
+    def add_packages(pkg_list)
+      @pkg_list.concat(pkg_list)
+    end
+
+    # Add the packages for the storage features in 'devicegraph' to the list of
+    # packages to be installed
+    #
+    # @param [::Storage::devicegraph] devicegraph to obtain the features from
+    # @return [Array<String>] new package list (may contain duplicates)
+    #
+    def add_feature_packages(devicegraph)
+      used_features = UsedStorageFeatures.new(devicegraph)
+      add_packages(used_features.feature_packages)
+    end
+
+    # Start a package dependency resolver run
+    #
+    def solve
+      Yast::Pkg.PkgSolve(true)
+    end
+
+    # Commit the changes depending on the current mode: During OS installation,
+    # mark the package list for installation. In the installed system, install
+    # them immediately.
+    #
+    def commit
+      if Yast::Mode.installation
+        set_proposal_packages
+      else
+        install!
+      end
+    end
+
+    # Execute package installation. This will install the stored package list
+    # immediately, so it is not advisable to do this during the OS
+    # installation. In the latter case, use 'set_proposal_packages' instead.
+    #
+    # @return true on success, false on error
+    #
+    def install!
+      compact
+      return if @pkg_list.empty?
+      log.info("Installing #{pkg_list}")
+      success = Yast::Package.DoInstall(@pkg_list)
+      if !success
+        log.error("ERROR: Some packages could not be installed")
+        install_error_popup
+      end
+      success
+    end
+
+    # Set the proposal packages for storage. This marks the packages for
+    # installation; it does not install them yet.
+    #
+    def set_proposal_packages
+      compact
+      return if @pkg_list.empty?
+      log.info("Marking #{pkg_list} for installation")
+      if !Yast::PackagesProposal.SetResolvables(PROPOSAL_ID, :package, @pkg_list)
+        log.error("PackagesProposal::SetResolvables() for #{pkg_list} failed")
+        set_resolvables_error_popup
+      end
+      solve
+      nil
+    end
+
+    # Remove duplicates from the package list and those packages that are
+    # already installed.
+    #
+    # @return [Array<String>] compacted list
+    #
+    def compact
```

Comment:
If the application code chooses to do stupid things in the meantime, the result will be surprising. That's a normal fact of life. The life time of an instance of this class is not meant to stretch over all kinds of weird things. There is more that could change in the mean time: The device graph might change, the installed packages on the system might change, whatever. If you want to be extra sure, you will simply have to call `add_feature_packages` again.

Calling any of `commit`, `install`, `set_proposal_packages` will implicitly call `compact` and even handle such cases as packages have already been installed, so an `install` followed by an `install` will do no harm.

Having `compact` available to the public will also do no harm; it just removes packages that not or no longer need to be installed. It's the defensive approach, and that defensive approach is not subject to negotiation from my side.